ulfius: transitive dependencies are not visible to CMake

### Nixpkgs version

- Unstable (26.05)

### Describe the bug

When configuring a C CMake project with `ulfius` as a dependency, it fails to configure due to missing transitive dependencies.

I needed to add all of the following packages to `propagatedBuildInputs` in order to compile a simple CMake project with ulfius as a dependency.
```nix
pkgs.ulfius.overrideAttrs (final: prev: {
  propagatedBuildInputs = with pkgs; [
    orcania
    yder
    libmicrohttpd
    gnutls
    jansson
    curl.dev
  ];
})
```

### Steps to reproduce

1. Have `ulfius` and `cmake` derivations installed.
2.  Create a basic C project with a `main.c` and `CMakeLists.txt` with `find_package(Ulfius REQUIRED)` and `target_link_libraries(my_target PRIVATE ulfius)`.
3. Try to configure it (e.g.: `cmake -S . -B build`)

### Expected behaviour

The CMake project should be configured without errors.

### Screenshots

_No response_

### Relevant log output

```console
CMake Error at /nix/store/wmb16s8hw1vd25vvidwnncpzq126lcz3-cmake-4.1.2/share/cmake-4.1/Modules/CMakeFindDependencyMacro.cmake:78 (find_package):
  By not providing "FindOrcania.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Orcania", but
  CMake did not find one.

  Could not find a package configuration file provided by "Orcania" with any
  of the following names:

    OrcaniaConfig.cmake
    orcania-config.cmake

  Add the installation prefix of "Orcania" to CMAKE_PREFIX_PATH or set
  "Orcania_DIR" to a directory containing one of the above files.  If
  "Orcania" provides a separate development package or SDK, be sure it has
  been installed.
Call Stack (most recent call first):
  /nix/store/9wqva4yr29zjzks26j4zspzvvlarjysa-ulfius-2.7.15/lib/cmake/Ulfius/UlfiusConfig.cmake:52 (find_dependency)
  src/jano_proxy/CMakeLists.txt:5 (find_package)
```

### Additional context

_No response_

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.62, NixOS, 26.05 (Yarara), 26.05.20251215.1306659`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.31.2+1`
 - nixpkgs: `/nix/store/nv0f6mcdixfwhjqxnnrk3iwfvkgbrxjl-source`

### Notify maintainers

@drupol 

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+label%3A%220.kind%3A+bug%22+-label%3A%226.topic%3A+darwin%22+-label%3A%226.topic%3A+nixos%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/me/meshtasticd/package.nix
+++ b/pkgs/by-name/me/meshtasticd/package.nix
@@ -16,13 +16,6 @@
   libuv,
   libxkbcommon,
   ulfius,
-  openssl,
-  gnutls,
-  jansson,
-  zlib,
-  libmicrohttpd,
-  orcania,
-  yder,
   yaml-cpp,
   udevCheckHook,
   versionCheckHook,
@@ -73,22 +66,15 @@ stdenv.mkDerivation (finalAttrs: {
 
   buildInputs = [
     bluez
-    gnutls
     i2c-tools
-    jansson
     libX11
     libgpiod_1
     libinput
-    libmicrohttpd
     libusb1
     libuv
     libxkbcommon
-    openssl
-    orcania
     ulfius
     yaml-cpp
-    yder
-    zlib
   ];
 
   preConfigure = ''

--- a/pkgs/by-name/ul/ulfius/package.nix
+++ b/pkgs/by-name/ul/ulfius/package.nix
@@ -2,7 +2,18 @@
   lib,
   stdenv,
   fetchFromGitHub,
+
+  withJson ? true,
+  withHttps ? true,
+  withWebsockets ? true,
+  withCurl ? true,
+  withLogger ? true,
+  withUwsc ? withWebsockets, # uwsc depends on websockets
+
+  # nativeBuildInputs
   cmake,
+
+  # Optional dependencies
   curl,
   gnutls,
   jansson,
@@ -29,14 +40,24 @@ stdenv.mkDerivation (finalAttrs: {
     cmake
   ];
 
-  buildInputs = [
-    curl
-    gnutls
-    jansson
+  propagatedBuildInputs = [
     libmicrohttpd
     orcania
-    yder
-    zlib
+  ]
+  ++ lib.optionals withJson [ jansson ]
+  ++ lib.optionals withCurl [ curl ]
+  ++ lib.optionals (withHttps || withWebsockets) [ gnutls ]
+  ++ lib.optionals withLogger [ yder ]
+  ++ lib.optionals withWebsockets [ zlib ];
+
+  cmakeFlags = [
+    (lib.cmakeBool "WITH_CURL" withCurl)
+    (lib.cmakeBool "WITH_JANSSON" withJson)
+    (lib.cmakeBool "WITH_GNUTLS" (withHttps || withWebsockets))
+    (lib.cmakeBool "WITH_YDER" withLogger)
+    (lib.cmakeBool "BUILD_UWSC" (withUwsc && withWebsockets))
+    (lib.cmakeBool "WITH_WEBSOCKET" withWebsockets)
+    (lib.cmakeBool "WITH_WEBSOCKET_MESSAGE_LIST" withWebsockets)
   ];
 
   meta = {
@@ -45,7 +66,7 @@ stdenv.mkDerivation (finalAttrs: {
     changelog = "https://github.com/babelouest/ulfius/blob/${finalAttrs.src.tag}/CHANGELOG.md";
     license = lib.licenses.lgpl21Only;
     maintainers = with lib.maintainers; [ drupol ];
-    mainProgram = "uwsc";
     platforms = lib.platforms.all;
-  };
+  }
+  // lib.optionalAttrs (withUwsc && withWebsockets) { mainProgram = "uwsc"; };
 })

